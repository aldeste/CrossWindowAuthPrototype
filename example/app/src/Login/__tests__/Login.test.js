import React from "react";
import Login from "../Login";
import renderer from "react-test-renderer";
require("jasmine-check").install();

jest.mock("../../Tags");

const PostWindowMessage = jest.fn();
global.window = { postMessage: msg => PostWindowMessage(msg) };
global.window.top = { postMessage: msg => PostWindowMessage(msg) };
global.document = { querySelector: el => ({}) };

describe("Login react component", () => {
  it("Is a react component", () => {
    expect(typeof Login).toBe("function");
    expect(<Login />.$$typeof.toString()).toBe("Symbol(react.element)");
  });

  it("Matches earlier configuration", () => {
    const tree = renderer
      .create(<Login title="This is a title" prefix="formName" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Enterint text in input changes value", () => {
    const component = renderer.create(<Login prefix="formName" />);
    ["username", "password"].forEach(name =>
      component
        .getInstance()
        .handleInputChange({ target: { name, value: "haschanged" } })
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Marks button as valid if both fields has 5 characters", () => {
    const component = renderer.create(<Login prefix="formName" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    ["username", "password"].forEach(name =>
      component
        .getInstance()
        .handleInputChange({ target: { name, value: "More than 5" } })
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  check.it("Accepts string values in input fields", gen.string, value => {
    const component = renderer.create(<Login prefix="formName" />);
    const fields = ["username", "password"];

    fields.forEach(name =>
      component.getInstance().handleInputChange({ target: { name, value } })
    );
    fields.forEach(name =>
      expect(typeof component.getInstance().state[name]).toBe("string")
    );
  });

  check.it("can't start with a whitespace", gen.string, randomString => {
    const component = renderer.create(<Login prefix="formName" />);
    component
      .getInstance()
      .handleInputChange({ target: { name: "username", value: randomString } });
    expect(component.getInstance().state.username.match(/^\s+/)).toBeFalsy();
  });

  check.it("can't have double whitespace strings", gen.string, randomString => {
    const component = renderer.create(<Login prefix="formName" />);
    component
      .getInstance()
      .handleInputChange({ target: { name: "username", value: randomString } });
    expect(component.getInstance().state.username).not.toContain("  ");
  });

  it("Returns an error if username is invalid", async () => {
    global.fetch = () =>
      new Promise(resolve =>
        resolve({ json: () => ({ error: "This is an error" }) })
      );

    const component = renderer.create(
      <Login title="This is a title" onLoginSubmit={jest.fn} />
    );

    let tree = component.toJSON();

    component.getInstance().handleInputChange({
      target: {
        name: "username",
        value: "This name doesn't exist in Star Wars"
      }
    });
    component
      .getInstance()
      .handleInputChange({ target: { name: "password", value: "password" } });
    await component
      .getInstance()
      .handleOnSubmit({ preventDefault: jest.fn, stopPropagation: jest.fn });

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Returns an error if password is invalid", async () => {
    global.fetch = () =>
      new Promise(resolve =>
        resolve({ json: () => ({ error: "This is an error" }) })
      );

    const component = renderer.create(
      <Login
        title="This is a title"
        prefix="formName"
        onLoginSubmit={jest.fn}
      />
    );
    component
      .getInstance()
      .handleInputChange({ target: { name: "username", value: "Yoda" } });
    component.getInstance().handleInputChange({
      target: { name: "password", value: "Wrong password this is" }
    });
    await component
      .getInstance()
      .handleOnSubmit({ preventDefault: jest.fn, stopPropagation: jest.fn });
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Signs in the user of password and username are correct", async () => {
    global.fetch = () =>
      new Promise(resolve =>
        resolve({
          json: () => ({
            user: "This is an error"
          })
        })
      );

    const component = renderer.create(
      <Login
        title="This is a title"
        prefix="formName"
        onLoginSubmit={jest.fn}
      />
    );
    component
      .getInstance()
      .handleInputChange({ target: { name: "username", value: "Yoda" } });
    component
      .getInstance()
      .handleInputChange({ target: { name: "password", value: "password" } });
    await component
      .getInstance()
      .handleOnSubmit({ preventDefault: jest.fn, stopPropagation: jest.fn });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Calls the onLoginSubmit callback on succsessfull loggin", async () => {
    global.fetch = () =>
      new Promise(resolve =>
        resolve({
          json: () => ({
            user: "This isn't an error"
          })
        })
      );

    const onLoginSubmit = jest.fn();
    const component = renderer.create(
      <Login
        title="This is a title"
        prefix="formName"
        onLoginSubmit={onLoginSubmit}
      />
    );

    component
      .getInstance()
      .handleInputChange({ target: { name: "username", value: "Yoda" } });
    component
      .getInstance()
      .handleInputChange({ target: { name: "password", value: "password" } });
    await component
      .getInstance()
      .handleOnSubmit({ preventDefault: jest.fn, stopPropagation: jest.fn });
    expect(onLoginSubmit).toHaveBeenCalled();
  });

  it("Doesn't calls the onLoginSubmit callback on error", async () => {
    global.fetch = () =>
      new Promise(resolve =>
        resolve({
          json: () => ({
            error: "This is an error"
          })
        })
      );

    const onLoginSubmit = jest.fn();
    const component = renderer.create(
      <Login
        title="This is a title"
        prefix="formName"
        onLoginSubmit={onLoginSubmit}
      />
    );

    component
      .getInstance()
      .handleInputChange({ target: { name: "username", value: "Yoda" } });
    component.getInstance().handleInputChange({
      target: { name: "password", value: "Wrong password this is" }
    });
    await component
      .getInstance()
      .handleOnSubmit({ preventDefault: jest.fn, stopPropagation: jest.fn });
    expect(onLoginSubmit).not.toHaveBeenCalled();
  });
});
