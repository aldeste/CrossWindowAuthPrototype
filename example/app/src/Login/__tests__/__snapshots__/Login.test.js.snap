// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login react component Enterint text in input changes value 1`] = `
<div
  className="View__View-jelnfZ bnYeJt"
>
  <h1
    className="Title__Title-lbDhID gxinMP"
  />
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    Please log in
  </p>
  <form
    className="Form__Form-PASBS imuxEL"
    onSubmit={[Function]}
  >
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNameusername"
        required={false}
      >
        Username
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNameusername"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        value="haschanged"
      />
    </div>
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNamepassword"
        required={false}
      >
        Password
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNamepassword"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value="haschanged"
      />
    </div>
    <button
      className="Button__Button-jkHDFw jklsEJ"
      disabled={false}
    >
      Sign in
    </button>
  </form>
</div>
`;

exports[`Login react component Marks button as valid if both fields has 5 characters 1`] = `
<div
  className="View__View-jelnfZ bnYeJt"
>
  <h1
    className="Title__Title-lbDhID gxinMP"
  />
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    Please log in
  </p>
  <form
    className="Form__Form-PASBS imuxEL"
    onSubmit={[Function]}
  >
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNameusername"
        required={false}
      >
        Username
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNameusername"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        value=""
      />
    </div>
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNamepassword"
        required={false}
      >
        Password
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNamepassword"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <button
      className="Button__Button-jkHDFw jklsEJ"
      disabled={true}
    >
      Form fields not valid
    </button>
  </form>
</div>
`;

exports[`Login react component Marks button as valid if both fields has 5 characters 2`] = `
<div
  className="View__View-jelnfZ bnYeJt"
>
  <h1
    className="Title__Title-lbDhID gxinMP"
  />
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    Please log in
  </p>
  <form
    className="Form__Form-PASBS imuxEL"
    onSubmit={[Function]}
  >
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNameusername"
        required={false}
      >
        Username
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNameusername"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        value="More than 5"
      />
    </div>
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNamepassword"
        required={false}
      >
        Password
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNamepassword"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value="More than 5"
      />
    </div>
    <button
      className="Button__Button-jkHDFw jklsEJ"
      disabled={false}
    >
      Sign in
    </button>
  </form>
</div>
`;

exports[`Login react component Matches earlier configuration 1`] = `
<div
  className="View__View-jelnfZ bnYeJt"
>
  <h1
    className="Title__Title-lbDhID gxinMP"
  >
    This is a title
  </h1>
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    Please log in
  </p>
  <form
    className="Form__Form-PASBS imuxEL"
    onSubmit={[Function]}
  >
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNameusername"
        required={false}
      >
        Username
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNameusername"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        value=""
      />
    </div>
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNamepassword"
        required={false}
      >
        Password
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNamepassword"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <button
      className="Button__Button-jkHDFw jklsEJ"
      disabled={true}
    >
      Form fields not valid
    </button>
  </form>
</div>
`;

exports[`Login react component Returns an error if password is invalid 1`] = `
<div
  className="View__View-jelnfZ bnYeJt"
>
  <h1
    className="Title__Title-lbDhID gxinMP"
  >
    This is a title
  </h1>
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    Please log in
  </p>
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    It appears something went terribly wrong!
  </p>
  <form
    className="Form__Form-PASBS imuxEL"
    onSubmit={[Function]}
  >
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNameusername"
        required={false}
      >
        Username
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNameusername"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        value="Yoda"
      />
    </div>
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNamepassword"
        required={false}
      >
        Password
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNamepassword"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value="Wrong password this is"
      />
    </div>
    <button
      className="Button__Button-jkHDFw jklsEJ"
      disabled={false}
    >
      Sign in
    </button>
  </form>
</div>
`;

exports[`Login react component Returns an error if username is invalid 1`] = `
<div
  className="View__View-jelnfZ bnYeJt"
>
  <h1
    className="Title__Title-lbDhID gxinMP"
  >
    This is a title
  </h1>
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    Please log in
  </p>
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    It appears something went terribly wrong!
  </p>
  <form
    className="Form__Form-PASBS imuxEL"
    onSubmit={[Function]}
  >
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="undefinedusername"
        required={false}
      >
        Username
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="undefinedusername"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        value="This name doesn't exist in Star Wars"
      />
    </div>
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="undefinedpassword"
        required={false}
      >
        Password
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="undefinedpassword"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value="password"
      />
    </div>
    <button
      className="Button__Button-jkHDFw jklsEJ"
      disabled={false}
    >
      Sign in
    </button>
  </form>
</div>
`;

exports[`Login react component Signs in the user of password and username are correct 1`] = `
<div
  className="View__View-jelnfZ bnYeJt"
>
  <h1
    className="Title__Title-lbDhID gxinMP"
  >
    This is a title
  </h1>
  <p
    className="Text__Text-jpiXUO hvuOVR"
  >
    Please log in
  </p>
  <form
    className="Form__Form-PASBS imuxEL"
    onSubmit={[Function]}
  >
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNameusername"
        required={false}
      >
        Username
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNameusername"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        value="Yoda"
      />
    </div>
    <div
      className="View__View-jelnfZ bnYeJt"
    >
      <label
        className="Label__Label-jsURsk iCwAoZ"
        htmlFor="formNamepassword"
        required={false}
      >
        Password
      </label>
      <input
        className="TextInput__Text-jwTuSF bRSPUO"
        id="formNamepassword"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value="password"
      />
    </div>
    <button
      className="Button__Button-jkHDFw jklsEJ"
      disabled={false}
    >
      Sign in
    </button>
  </form>
</div>
`;
